sycl/doc/KernelProgramCache.md

```diff
@@ -0,0 +1,223 @@
+execution. Cacheability of an object can be tested with
+`program_impl::is_cacheable()` method. SYCL RT will only try to insert cacheable
+programs or kernels into the cache. This is done as a part of
+`ProgramManager::getOrCreateKernel()` method.
+
+
+*NOTE:* a kernel is only cacheable if and only if the program it belongs to is
+cacheable. On the other hand if the program is cacheable, then each and every
+kernel of this program will be cached also.
+
+
+All requests to build a program or to create a kernel - whether they originate
+from explicit user API calls or from internal SYCL runtime execution logic - end
+up with calling the function [`getOrBuild()`](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/program_manager/program_manager.cpp#L149)
+with number of lambda functions passed as arguments:
+ - Acquire function;
+ - GetCache function;
+ - Build function.
+
+*Acquire* function returns a locked version of cache. Locking is employed for
+thread safety. The threads are blocked only for insert-or-acquire attempt, i.e.
+when calling to `map::insert` in [`getOrBuild`](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/program_manager/program_manager.cpp#L149)
+function. The rest of operation is done with the help of atomics and condition
+variables (plus a mutex for proper work of condition variable).
+
+*GetCache* function returns a reference to mapping `key->value` out of locked
+instance of cache. We will see rationale behind it a bit later.
+
+*Build* function actually builds the kernel or program.
+
+Caching isn't done:
+ - when program is built out of source with `program::build_with_source()` or
+   `program::compile_with_source()` method;
+ - when program is a result of linking multiple programs.
+
+
+### Thread-safety
+
+Why do we need thread safety here? It is quite possible to have a use-case when
+the `cl::sycl::context` is shared across multiple threads (e.g. via sharing a
+queue). Possibility of enqueueing multiple cacheable kernels simultaneously
+from multiple threads requires us to provide thread-safety for the caching
+mechanisms.
+
+It is worth of noting that we don't cache the PI resource (kernel or program)
+by itself. Instead we augment the resource with the status of build process.
+Hence, what is cached is a wrapper structure `BuildResult` which contains three
+information fields - pointer to built resource, build error (if applicable) and
+current build status (either of "in progress", "succeeded", "failed").
+
+One can find definition of `BuildResult` template in [KernelProgramCache](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/kernel_program_cache.hpp).
+
+Pointer to built resource and build result are both atomic variables. Atomicity
+of these variables allows one to hold lock on cache for quite a short time and
+perform the rest of build/wait process without forcing other threads to wait on
+lock availability.
```

Comment:
I think it is not the atomicity of fields which allows that, but the synchronization approach 
used. In this strategy fields in the BuildResult are accessed from multiple threads, so they must be atomic to ensure proper memory fences and exclude visibility of partially written data.

```suggestion
The built resource access synchronization approach aims at minimizing the time any thread 
holds the global lock guarding the maps to improve performance. To achieve that, the global lock is acquired only for the duration of the global map access. Actual build of the program happens outside of the lock, so other threads can request or build other programs in the meantime. A thread requesting a `BuildResult` instance via `getOrBuild` can go one of three ways:
A) Build result is **not** available, it is the first thread to request it. Current thread will then execute the build letting others wait for the result using the per-build result condition variable kept in `BuildResult`'s `MBuildCV` field.
B) Build result is **not** available, another thread is already building the result. Current thread will then wait for the result using the `MBuildCV` condition variable.
C) Build result **is** available. The thread simply takes it from the `Ptr` field w/o using any mutexes or condition variables.
As noted before, access to `BuildResult` instance fields may occur from different threads simultaneously, but the global lock is no longer held. So, to make it safe and to make sure only one thread builds the requested program, the following is done:
- program build state is reflected in the `State` field, threads use compare-and-swap technique to compete who will do the build and become thread A. Threads C will find 'DONE' in this field and immediately return the with built result at hand.
- thread A and thread(s) B use the `MBuildCV` conditional variable field and `MBuildResultMutex`  mutex field guarding that variable to implement the "single producer-multiple consumers scheme". 
- the build result itself appears in the 'Ptr' field when available.
All fields are atomic because they can accessed from multiple threads.
```

Reply:
Good point. Changed.